### PR TITLE
Added template validation

### PR DESCRIPTION
**Pull request by itavy**
if it is ok i will add a test for this case

Signed-off-by: Octavian Ionescu <itavyg@gmail.com>

### CODE REVIEW

File: azurerm/resource_arm_template_deployment.go [hunk 1]
```diff
@@ -152,6 +155,22 @@ func resourceArmTemplateDeploymentCreateUpdate(d *schema.ResourceData, meta inte
...
+
+	if err != nil {
+		return fmt.Errorf("Error on communicating with portal to validate Template for Deployment: %+v", err)
+	}
+
+	if deploymentValidateResponse.Error != nil {
+		if deploymentValidateResponse.Error.Message != nil {
+			return fmt.Errorf("Error validating Template for Deployment: %+v", deploymentValidateResponse.Error.Message)
+		}
+		return fmt.Errorf("Error validating Template for Deployment: %+v", deploymentValidateResponse.Error)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_template_deployment.go]: could we expose the name of the deployment/resource group here? we do this in other resources to show which deployment failed if users have multiple:

```suggestion
		return fmt.Errorf("Error validating Template for Deployment %q (Resource Group %q): %+v", name, resourceGroup, *deploymentValidateResponse.Error)
```

(we also need to de-reference this)

File: azurerm/resource_arm_template_deployment.go [hunk 2]
```diff
@@ -152,6 +155,22 @@ func resourceArmTemplateDeploymentCreateUpdate(d *schema.ResourceData, meta inte
...
 
+	deploymentValidateResponse, err := client.Validate(ctx, resourceGroup, name, deployment)
+
+	if err != nil {
+		return fmt.Errorf("Error on communicating with portal to validate Template for Deployment: %+v", err)
+	}
+
+	if deploymentValidateResponse.Error != nil {
+		if deploymentValidateResponse.Error.Message != nil {
+			return fmt.Errorf("Error validating Template for Deployment: %+v", deploymentValidateResponse.Error.Message)
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_template_deployment.go]:
```suggestion
			return fmt.Errorf("Error validating Template for Deployment %q (Resource Group %q): %+v", name, resourceGroup, *deploymentValidateResponse.Error.Message)
```

(we also need to dereference this)

File: azurerm/resource_arm_template_deployment.go [hunk 3]
```diff
@@ -97,6 +97,9 @@ func resourceArmTemplateDeploymentCreateUpdate(d *schema.ResourceData, meta inte
 	resourceGroup := d.Get("resource_group_name").(string)
 	deploymentMode := d.Get("deployment_mode").(string)
 
+	// following instructions won't make any change in azure
+	// so they shouldn't change the state
+	d.Partial(true)
```
[the remark this answers is not in the document]
Comment (itavy) [reply azurerm/resource_arm_template_deployment.go]: it actually change the state
i tried with an invalid template:
```
$ terraform plan
...
Plan: 0 to add, 1 to change, 0 to destroy.
$ terraform apply
......
Error: Error validating Template for Deployment "****" (Resource Group "****"):  Deployment template validation failed: 'The value for the template parameter 'api_key' at line '1' and column '227' is not provided. Please see https://aka.ms/resource-manager-parameter-files for usage details.'.
$ terraform plan
...
No changes. Infrastructure is up-to-date.
```

[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_template_deployment.go]: potentially during update, in which case these'd want `if !d.IsNewResource() { .. } ` wrapped around this to only occur during an update

**Comment by tombuildsstuff** [comment]
hey @itavy 

I've run the tests for this and aside from one existing failing test this LGTM:

Screenshot 2019-10-29 at 12 55 52

I hope you don't mind but to make that final test pass I've pushed a commit to this branch which changes the `TestAccAzureRMTemplateDeployment_withParamsBody` test from a Complete to an Incremental deployment, which whilst outside the scope of this PR makes these tests all pass:

Screenshot 2019-10-29 at 13 23 36

Thanks!

**Pull request merged**